### ISSUE TITLE
(1.7.1 iOS) HP slider issues

### ISSUE THREAD

**Issue by SaschB** [Tarmslitaren/FrosthavenAssistant#91]
While I appreciate the effort of adding HP slide adjustments, here are some things I noticed from first exposure, ranging from bugs to preferences:

* Breaks on Undo: the values available in the slider overlay stop updating after performing undo of health change, making it possible to go above max or below min health, as well as forcing health changes to be made in increments based on where it got stuck

* Character box does not regain color if HP is increased from 0 (not a common scenario, but possible by mistake, and Undo leads to other issues)

* Drag sensitivity appears too high, at least for me, it's hard to hit the desired value without focusing. This increases the risk for accidentally setting HP to zero, leading into the above issues

* Target area for start of slide adjustment is very small, making it easy to fail to initiate. Suggestion: allow drag action to start from anywhere on the character box (like in GHH)

* Due to the small target area and high scroll speed, once the scroll has been successfully started, it's easy to have already passed the target value, having to precisely maneuver backwards

* While sliding the wheel, the HP doesn't update to reflect the result. This might be intentional, but forces the math onto the user when there is a target health ("Please heal this figure to 17")

* Decreasing HP requires sliding to the right (and vice versa). This might be intentional, but felt unintuitive to me

These last two points make me believe the feature was designed from a standpoint of "I want to add X damage to this figure", but it's quite hard to perform the action "Set this figure's HP to 9" as a result.

The GHH approach with live feedback in the character bar while sliding, but not committing the change until the pointer is released would be very good as an option for users who prefer the later design.

**Comment by Tarmslitaren** [comment]
the sensitivity is depending on max health so you'd be able to see more of the range on high health monsters. I can tweak it it be less sensitive in general though.

I guess it's somewhat counterintuitive to drag right to decrease. One reason is, unlike ghh the monster standees go all the way to the edge, so swiping left is hard to do, and the more common operation would be to decrease health.

The typical usecase would be to 'heal 7' or 'damage 3'. I would think it's rare to see a 'heal to exactly 17' situation? ( to be fair, the reason for not updating the health value before committing is due to how much I'd need to change the code )

**Comment by SaschB** [comment]
I didn't play with it a lot, but it happened to be with < 10 max hp.

Don't the standees also go all the way to the right though? Not as often, maybe, but some enemies have many standees in play at the same time.

I can see the logic behind focusing on the difference in health to apply. GHH did show both values live, so whichever you were interested in was always at hand. I guess people use it in various ways.

The main issue here was the undo functionality, and a bit about the usability, the later points were mostly observations.

**Comment by Tarmslitaren** [comment]
I am actually very grateful for the observations ( especially the bug report, very valuable)

I think, if you see it as a wheel ( which, in design it is ) it makes sense to draw right to turn the wheel left, and I'd say it makes sense for negative values to be on the left side. It's probably not too hard to get used to.

I guess I could add another row of values showing the absolute health values as well, if that helps, and is not too confusing.

I never used ghh drag functionality, but testing it out it seemed very slow to me, but I might have gone too far in the other direction, though having the 'wheel' visible would make it so you can easier fine tune where to stop, and so allow for faster movement.

**Comment by SaschB** [comment]
I did sometimes have issues with GHH requiring multiple swipes for large amounts. Not saying that was perfect either. The full width of my screen represented about 20 HP. Dynamic scaling seems useful for that!

Allowing the drag to start anywhere on the bar would do wonders for usability though (as long as no other action is tied to horizontal drag).

**Comment by GarrettChan0** [comment]
Don't want to create another issue since it's similar, so I may as well leave something here:

I also think the sensitivity is a bit high. I used GHH before and I think GHH is pretty comfortable when I was using it.

**Comment by Tarmslitaren** [comment]
I will absolutely add so you can drag from anywhere on character. I'll tweak the sensitivity to be a bit lower.

Considering how useful dynamically Scaling sensitivity is, since if it stays the same you can use your muscle memory to faster set common values, but then, for the cases where you want to change bigger values will be less usable.

Also need to remember to have a reasonable experience on all different type of screens, and input methods, it's not entirely straight forward. this whole feature requires non standard ux, so there's no right answer to find anywhere else.

**Comment by mtelequin** [comment]
Hello!
Loving the app and the recent release. I've been trying it out with my wife on an HP Spectre, Pixel, and iPhone. If we can buy @Tarmslitaren a coffee, let us know. We are big fans <3

In addition @SaschB's comments above, I did experience one bug where the slider allowed me to over-deplete the hitpoints on a monster. This caused the dying animation to occur, where the hp tracker for the monster slides below the card. However, the tracker froze at the bottom of its animation track with a negative value and never went away until I removed that monster set entirely from the scenario. I've attached an example.



![image](https://user-images.githubusercontent.com/125409048/218867095-859fcee7-85c3-4ccd-ac8e-b5b1ffd6eaa9.png)

**Comment by SaschB** [comment]
> In addition @SaschB's comments above, I did experience one bug where the slider allowed me to over-deplete the hitpoints on a monster. 

Do you know if you used the Undo functionality at all, or was this a separate way of reaching negative HP?

**Comment by Tarmslitaren** [comment]
It's definitely a ui update issue. for some reason the overlay widgets don't get updated on changes not originating on the character/ monster box. 
a workaround is to long press on the monster/character and it resets.

**Comment by mtelequin** [comment]
> > In addition @SaschB's comments above, I did experience one bug where the slider allowed me to over-deplete the hitpoints on a monster.
> 
> Do you know if you used the Undo functionality at all, or was this a separate way of reaching negative HP?

I did not use the undo function in this particular instance.



> It's definitely a ui update issue. for some reason the overlay widgets don't get updated on changes not originating on the character/ monster box. a workaround is to long press on the monster/character and it resets.

This is super helpful. Thanks!

**Comment by Tarmslitaren** [comment]
About the sensitivity: what platforms are you using?
I tested only on Android and Windows while developing the feature, and found those to work decently well ( will tweak maybe a little bit ). Was checking now on an ipad and the sensitivity is clearly too high on that one, so it could potentially be a platform dependant problem

**Comment by Tarmslitaren** [comment]
All of the mentioned issues, except showing absolute health values, have been addressed in version 1.7.2, closing this now. If there are wishes for further changes, or new issues found, please open a new ticket.

**Comment by SaschB** [comment]
> About the sensitivity: what platforms are you using?

I'm on iOS (iPhone and iPad), we played a few scenarios yesterday (on 1.7.1), and I often ended up slightly overshooting the target value (e.g. landing on -5 instead of -4), and having to be very precise to adjust it back to the target (-4) without overshooting again (landing on -3). Could very well be OS dependent, I'll check again on 1.7.2.

Other than that, and some of the issues that have been fixed, it worked pretty well. I noticed that I only do the manual tapping health adjustment if there is also an additional effect like a curse or status.
